pkg/operator/mysql/mysql.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2024 PlanetScale Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package mysql
+
+import (
+	"fmt"
+	"regexp"
+	"slices"
+	"strconv"
+	"strings"
+
+	"planetscale.dev/vitess-operator/pkg/operator/environment"
+	"planetscale.dev/vitess-operator/pkg/operator/vitess"
+	"vitess.io/vitess/go/vt/sqlparser"
+)
+
+var imageVersionRegExp = regexp.MustCompile(`^(\d+)\.(\d+)\.(\d+)`)
+
+func UpdateMySQLServerVersion(flags vitess.Flags, mysqldImage string) {
+	value, err := dockerImageGetVersionToString(mysqldImage)
+	if err != nil {
+		return
+	}
+	flags["mysql_server_version"] = value
+	environment.MySQLServerVersion = value
+}
+
+func dockerImageGetVersionToString(currentVersionImage string) (string, error) {
+	currentVersionSlice := strings.SplitN(currentVersionImage, ":", 2)
+	if len(currentVersionSlice) != 2 {
+		return "", fmt.Errorf("could not parse the image name and image label, got: %s, but expected xx:xx", currentVersionImage)
+	}
+
+	label := currentVersionSlice[1]
+	_, err := sqlparser.ConvertMySQLVersionToCommentVersion(label)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s-Vitess", label), nil
+}
+
+func DockerImageSafeUpgrade(currentVersionImage, desiredVersionImage string) (bool, error) {
+	if currentVersionImage == "" || desiredVersionImage == "" {
+		// No action if we have unknown versions.
```

Comment:
Decided to move this from `pkg/controller/vitessshardreplication/reconcile_drain.go` to this new package to ease readability, no change were applied to `DockerImageSafeUpgrade` besides renaming the function from `safeMysqldUpgrade` to this